icechunk/src/format/snapshot.rs

```diff
@@ -49,6 +49,55 @@ pub struct ZarrArrayMetadata {
     pub dimension_names: Option<DimensionNames>,
 }
 
+impl ZarrArrayMetadata {
+
+    /// Returns an iterator over the maximum permitted chunk indices for the array.
+    ///
+    /// This function calculates the maximum chunk indices based on the shape of the array
+    /// and the chunk shape.
+    ///
+    /// # Returns
+    ///
+    /// An iterator over the maximum permitted chunk indices.
+    fn max_chunk_indices_permitted(&self) -> impl Iterator<Item = u64> + '_ {
+        self.shape
+            .iter()
+            .zip(self.chunk_shape.0.iter())
+            .map(|(s, cs)| ((s + cs.get() - 1)) / cs.get() - 1)
```

Comment:
The math looks a bit weird to me. Shouldn't it be just `s / cs.get() - 1`?

Reply:
Yes I don't know why I left this expanded like this. Simplified it down to `(s - 1) / cs.get()`, which should be equivalent and also seems to handle all cases properly. 

I added tests to confirm this, but if there are additional edge cases I haven't thought of, let me know. Will continue to think on it.